### ISSUE TITLE
global limit on the num of running ignoreLocality jobs

### ISSUE THREAD

**Issue by juztas** [dmwm/CRABServer#4856]
```
> (2) HTCondor 8.3.7 features improved concurrency limits.  I expect a
> pre-release goes into ITB this week or next.  This should allow CRAB3 devs
> to add a global limit on the number of running ignoreLocality jobs (and a
> rough per-site limit).
```
@bbockelm can you point to ticket or pull request in HTCondor?
https://hypernews.cern.ch/HyperNews/CMS/get/comp-ops/2382/1/2/2/2/2.html

**Comment by bbockelm** [comment]
The relevant ticket in HTCondor is gt5006; it's committed for 8.3.7 pre-releases.  This allows the ConcurrencyLimit job attribute to reference the machine ad.

First, we would need to add the list of local sites to the job ad.  Perhaps here:

https://github.com/dmwm/CRABServer/blob/master/src/python/TaskWorker/Actions/DagmanCreator.py#L92

Then, we could add the following to the schedd configuration file:

```
SUBMIT_ATTRS = $(SUBMIT_ATTRS), ConcurrencyLimit
ConcurrencyLimit = ifThenElse(stringListMember(GLIDEIN_CMSSite, CRAB_LocalSites), "", "ignoreLocality")
```

(some adjustment might be needed for syntax, testing, etc).  Then, any job using ignoreLocality landing at a non-local site would have to obey the ignoreLocality limit.

Finally, we would need to enable consumption policies on the glidein.  This implies a frontend change for Farrukh, but cannot be done in production until the collector is 8.3.7 due to gt4973 (scalability fix necessary for global pool scale).

If this all works, it would be a global limit.  We could do marginally better by having a per-site limit *only* for cases where CRAB_LocalSites is exactly one site.

**Comment by belforte** [comment]
AFAICT this is for the submission infrastructure group, not CRAB code. Correct ? @bbockelm @juztas ?
Then we should make sure it goes in Jame's list and close.

**Comment by juztas** [comment]
only one thing is needed to add local sites classad. All other are for submission infrastructure. We still are running 8.3.5 so I would keep it open not to forget to implement. Next week we will have 8.3.8

**Comment by belforte** [comment]
what's list of local sites ? the list of sites that crab would have put in DESIRED_Sites if user had not indicated ignore_locality ? I'd rather use a more meaningful attribute name like "CRAB_data_locations".

**Comment by bbockelm** [comment]
Correct - we need to have something along the lines of "CRAB_Data_Locations" so we can compute from the negotiator when ignoreLocality is in use.

Once that's in place, you still need to set the ConcurrencyLimit value from the CRAB3 side.

At that point, we can indeed close the ticket and bug the submit infrastructure group about applying a limit.

**Comment by mmascher** [comment]
About the name of the classad (`CRAB_Data_Locations`), here  https://github.com/dmwm/CRABServer/issues/4413 Eric was asking to add `DESIRED_CMSDatasetLocations`. Should we use that name?

@ericvaandering what `DESIRED_CMSDatasetLocations` is going to be used for?

**Comment by belforte** [comment]
I think use case in WMA is exactly the same: make data location available for HTC so to optimize scheduling decisions. Therefore we should use same name. I.e. one reads `DESIRED_CMSDatasetLocations` not as `places where I like to run` which is 'DESIRED_Sites` but as 'locations of the CMS Dataset I desire to use`

**Comment by mmascher** [comment]
At the end I decided to go with the name `CRAB_DataLocation` as it's not really the dataset location but more the block location (so data and not dataset)

Moreover the naming is consitent with the CRAB3 conventions and we already have a classad for the input datasets.

**Comment by ericvaandering** [comment]
Except the idea was to harmonize this between the Agent and CRAB so that any required GlideinWMS changes don't have to look for multiple values to do the same thing.

**Comment by mmascher** [comment]
Yeah, indeed I also thought about that and I had some doubts about what I did. Maybe we can change the name of the classad https://github.com/dmwm/WMCore/pull/5989#issuecomment-140842195 ?

Then changing the name of the classad `CRAB_InputData` which is used in the crab3 code would be more work. And again all the crab3 classads starts with `CRAB_`. Maybe it makes sense to keep both CRAB_InputData and DESIRED_CMSBlockLocations

**Comment by belforte** [comment]
*crab3 specific* classads start with CRAB ! But e.g. you do not put CRAB_DESIRED_Site
So please, use the same. I know that picking good names is difficult, but this is getting ridicolous :-)
And if changing CRAB code is more work than changing WMCore, it means to me that something is very 
very wrong in CRAB.

**Comment by ericvaandering** [comment]
It's not difficult either place. Especially since it's not used anywhere yet. :-)

**Comment by mmascher** [comment]
For the locations I changed it to use `DESIRED_CMSDataLocations`. Notice it's `Data` and not `Dataset`, let's change WMAgent as well (and I suggest to check you are getting block locations).

I renamed `CRAB_InputData` to `DESIRED_CMSDataset` . I considered simply adding `+DESIRED_CMSDataset = CRAB_InputData` to avoid problems in case this classad "leaked" outside (let's say it's used by DDM. ) But I xchecked with Niccolo' and it's not the case.
